### PR TITLE
Add ZipCode Counting Notebook

### PR DESCRIPTION
**Pull request by isVoid**
## Description

This PR adds a notebook that demonstrate the use of quadtree PiP with a custom `QuadTree` structure and joins dataframe. We can use this PR to discuss the possibility of adding the custom structure to the codebase.

This also updates the dependency list to include notebook environment in "all" targeted conda envs.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cuspatial/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by isVoid** [comment]
Pending update to Quadtree parameters.

File: notebooks/ZipCodes_Stops_PiP_cuSpatial.ipynb [hunk 1]
```diff
@@ -0,0 +1,754 @@
...
+   "cell_type": "markdown",
+   "id": "497810f3-acf0-4472-a187-322413c9db11",
+   "metadata": {},
+   "source": [
+    "## Overview\n",
+    "\n",
+    "Stop signs is a important symbol of city and road development in geographical information systems. This notebook processes all stop sign locations from the a dataset, using spatial joins to categorify them within the zipcodes boundaries that are located within California. This notebook does the following steps:\n",
+    "1. Filters the zipcode boundaries that locates in California with spatial join.\n",
+    "2. Filters the stop signs that locates in all the zipcodes with spatial join.\n",
+    "3. Count the stop signs by zipcodes.\n",
```
Comment (harrism) [inline notebooks/ZipCodes_Stops_PiP_cuSpatial.ipynb:41]:
```suggestion
    "Stop signs are an important symbol of city and road development in geographical information systems. This notebook processes all stop sign locations from the dataset, using spatial joins to locate them within the zipcode boundaries located within California. This notebook performs the following steps:\n",
    "1. Filters the zipcode boundaries located in California with spatial join.\n",
    "2. Filters the stop signs located in all the zipcodes with spatial join.\n",
    "3. Counts the stop signs by zipcode.\n",
```

File: notebooks/ZipCodes_Stops_PiP_cuSpatial.ipynb [hunk 2]
```diff
@@ -0,0 +1,754 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "044c84d9-2f82-4de5-a4c5-b7b9e5ef6b93",
+   "metadata": {},
+   "source": [
+    "## Load Dataset and Cleanup\n",
+    "\n",
+    "We load the dataset and store them as device dataframe in cuspatial. Note that the second cell below loads the dataset with cuDF, then adopts geopandas to parse the WKT strings into shapely objects. This is a slow step performed on CPU and requires data transfer between device and host."
```
Comment (harrism) [inline notebooks/ZipCodes_Stops_PiP_cuSpatial.ipynb:69]:
```suggestion
    "We load the datasets and store them as cuSpatial device dataframes. Note that the second cell below loads the dataset with cuDF, then adopts geopandas to parse the WKT (Well-Known Text) strings into shapely objects. This is a slow step performed on CPU and requires data transfer between device and host."
```

File: notebooks/ZipCodes_Stops_PiP_cuSpatial.ipynb [hunk 3]
```diff
@@ -0,0 +1,754 @@
...
+   "id": "583856a2-0b29-48c1-ade7-3840f640c5e7",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Load WKT as shapely objects\n",
+    "h_zip = d_zip.to_pandas()\n",
+    "h_zip[\"WKT\"] = h_zip[\"WKT\"].apply(wkt.loads)\n",
+    "h_zip = gpd.GeoDataFrame(h_zip, geometry=\"WKT\", crs='epsg:4326')\n",
+    "\n",
+    "# Roundtrip back to geopandas\n",
```
Comment (harrism) [inline notebooks/ZipCodes_Stops_PiP_cuSpatial.ipynb:110]:
```suggestion
    "# Transfer back to GPU with cuSpatial\n",
```

File: notebooks/ZipCodes_Stops_PiP_cuSpatial.ipynb [hunk 4]
```diff
@@ -0,0 +1,754 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "eda8fb4c-39ec-44e2-9163-cbbdd91eeb1d",
+   "metadata": {},
+   "source": [
+    "## Filtering Zipcode by its Geometric Center\n",
+    "\n",
+    "The Zipcode Dataset contains boundaries for all zipcodes in the US. The below uses the geometric center for each zipcode and uses cuspatial's quadtree interface to filter zipcodes that locates only in California."
```
Comment (harrism) [inline notebooks/ZipCodes_Stops_PiP_cuSpatial.ipynb:282]:
```suggestion
    "The Zipcode Dataset contains boundaries for all zipcodes in the US. The below uses the geometric center for each zipcode and uses cuspatial's quadtree interface to filter zipcodes located only in California."
```

File: notebooks/ZipCodes_Stops_PiP_cuSpatial.ipynb [hunk 5]
```diff
@@ -0,0 +1,754 @@
...
+     "text": [
+      "/raid/wangm/dev/rapids/cuspatial/python/cuspatial/cuspatial/core/spatial/indexing.py:174: UserWarning: scale -1 is less than required minimum scale 0.0029100948550503493. Clamping to minimum scale\n",
+      "  warnings.warn(\n",
+      "/raid/wangm/dev/rapids/cuspatial/python/cuspatial/cuspatial/core/spatial/join.py:171: UserWarning: scale -1 is less than required minimum scale 0.0029100948550503493. Clamping to minimum scale\n",
+      "  warnings.warn(\n"
+     ]
+    }
+   ],
+   "source": [
+    "# Build a 2nd quadtree with all stop signs in US\n",
```
Comment (harrism) [inline notebooks/ZipCodes_Stops_PiP_cuSpatial.ipynb:396]:
```suggestion
    "# Build a second quadtree with all stop signs in the US\n",
```

File: notebooks/ZipCodes_Stops_PiP_cuSpatial.ipynb [hunk 6]
```diff
@@ -0,0 +1,754 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "4b61f534-faa4-4465-b47b-fb717169f30e",
+   "metadata": {},
+   "source": [
+    "## Zipcode counting with cuDF\n",
+    "\n",
+    "The below resorts to [cuDF](https://docs.rapids.ai/api/cudf/stable/index.html) and counts the number of stop signs per zip code. Then merge the geometry information from the zipcode dataset."
```
Comment (harrism) [inline notebooks/ZipCodes_Stops_PiP_cuSpatial.ipynb:501]: "resorts to" has a negative connotation.
```suggestion
    "The below uses [cuDF](https://docs.rapids.ai/api/cudf/stable/index.html) to count the number of stop signs per zip code. Then merge the geometry information from the zipcode dataset."
```

**Review by harrism — CHANGES_REQUESTED** [review]
Just tidying up the text content.

File: notebooks/ZipCodes_Stops_PiP_cuSpatial.ipynb [hunk 7]
```diff
@@ -0,0 +1,754 @@
...
+  {
+   "cell_type": "markdown",
+   "id": "6931011f-0d83-45ce-b254-4b5424b82624",
+   "metadata": {},
+   "source": [
+    "This notebook demonstrates the use of geodataframe joins with point in polygon in cuSpatial\n",
+    "\n",
+    "## Prerequisite: Datasets\n",
+    "\n",
+    "Dataset used (TODO: discuss if need to self host):\n",
```
[the remark this answers is not in the document]
Comment (thomcom) [reply notebooks/ZipCodes_Stops_PiP_cuSpatial.ipynb:20]: I'm not inclined to self host, but it does have future regression implications to not do so.

[the remark this answers is not in the document]
Comment (isVoid) [reply notebooks/ZipCodes_Stops_PiP_cuSpatial.ipynb]: Hmm. I think the datasets used for this notebook is from OSM and processed into a csv file. I don't think they can be downloaded as is.

[the remark this answers is not in the document]
Comment (thomcom) [reply notebooks/ZipCodes_Stops_PiP_cuSpatial.ipynb]: Hey @taureandyernv are we clear to do automatic downloads from self hosted storage for these large datasets and run them in your testing framework?

**Review by thomcom — CHANGES_REQUESTED** [review]
Do we have a plan for getting the remote datasets yet? @taureandyernv did you see my post above?

**Comment by isVoid** [comment]
> Do we have a plan for getting the remote datasets yet? @taureandyernv did you see my post above?

Yes, @jarmak-nv has helped me to upload them to s3. But legal is still working through the authorization.

File: dependencies.yaml [hunk 8]
```diff
@@ -172,6 +172,7 @@ dependencies:
           - ipython
           - notebook
           - shapely
+          - pydeck
```
[the remark this answers is not in the document]
Comment (isVoid) [reply dependencies.yaml:176]: Added.

**Comment by isVoid** [comment]
@ajschmidt8 What's `pr / checks / check-generated-files` for and why does it fail?

File: notebooks/ZipCodes_Stops_PiP_cuSpatial.ipynb (hunk 7 above)
[the remark this answers is not in the document]
Comment (taureandyernv) [reply notebooks/ZipCodes_Stops_PiP_cuSpatial.ipynb]: yeah

**Comment by ajschmidt8** [comment]
> I will open up a separate PR to add this to `cuspatial` as a pre-commit hook similarly to [cudf](https://github.com/rapidsai/cudf/pull/12819), [cuml](https://github.com/rapidsai/cuml/pull/5246), etc.

PR opened below for reference:

- https://github.com/rapidsai/cuspatial/pull/1008

**Review by thomcom — APPROVED** [review]
Looking good.

**Pull request merged**